mempool/v1/reactor.go

```diff
@@ -0,0 +1,394 @@
+// object wraps a reference to a legacy p2p ChannelDescriptor and the corresponding
+// p2p proto.Message the new p2p Channel is responsible for handling.
+//
+//
+// TODO: Remove once p2p refactor is complete.
+// ref: https://github.com/tendermint/tendermint/issues/5670
+func GetChannelShims(config *cfg.MempoolConfig) map[p2p.ChannelID]*p2p.ChannelDescriptorShim {
+	largestTx := make([]byte, config.MaxTxBytes)
+	batchMsg := protomem.Message{
+		Sum: &protomem.Message_Txs{
+			Txs: &protomem.Txs{Txs: [][]byte{largestTx}},
+		},
+	}
+
+	return map[p2p.ChannelID]*p2p.ChannelDescriptorShim{
+		mempool.MempoolChannel: {
+			MsgType: new(protomem.Message),
+			Descriptor: &p2p.ChannelDescriptor{
+				ID:                  byte(mempool.MempoolChannel),
+				Priority:            5,
+				RecvMessageCapacity: batchMsg.Size(),
+
+				MaxSendBytes: 5000,
+			},
+		},
+	}
+}
+
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed.
+func (r *Reactor) OnStart() error {
+	if !r.config.Broadcast {
+		r.Logger.Info("tx broadcasting is disabled")
+	}
+
+	go r.processMempoolCh()
+	go r.processPeerUpdates()
+
+	return nil
+}
+
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit.
+func (r *Reactor) OnStop() {
+	r.mtx.Lock()
+	for _, c := range r.peerRoutines {
+		c.Close()
+	}
+	r.mtx.Unlock()
+
+	// wait for all spawned peer tx broadcasting goroutines to gracefully exit
+	r.peerWG.Wait()
+
+	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
+	// p2p Channels should execute Close().
+	close(r.closeCh)
+
+	// Wait for all p2p Channels to be closed before returning. This ensures we
+	// can easily reason about synchronization of all p2p Channels and ensure no
+	// panics will occur.
+	<-r.mempoolCh.Done()
+	<-r.peerUpdates.Done()
+}
+
+// handleMempoolMessage handles envelopes sent from peers on the MempoolChannel.
+// For every tx in the message, we execute CheckTx. It returns an error if an
+// empty set of txs are sent in an envelope or if we receive an unexpected
+// message type.
+func (r *Reactor) handleMempoolMessage(envelope p2p.Envelope) error {
+	logger := r.Logger.With("peer", envelope.From)
+
+	switch msg := envelope.Message.(type) {
+	case *protomem.Txs:
+		protoTxs := msg.GetTxs()
+		if len(protoTxs) == 0 {
+			return errors.New("empty txs received from peer")
+		}
+
+		txInfo := mempool.TxInfo{SenderID: r.ids.GetForPeer(envelope.From)}
+		if len(envelope.From) != 0 {
+			txInfo.SenderNodeID = envelope.From
+		}
+
+		for _, tx := range protoTxs {
+			if err := r.mempool.CheckTx(types.Tx(tx), nil, txInfo); err != nil {
+				logger.Error("checktx failed for tx", "tx", fmt.Sprintf("%X", mempool.TxHashFromBytes(tx)), "err", err)
```

Comment:
Is there ever an error when processing txs where we may want to send a peer update or peer error?

Reply:
The only case I can think of is the peer sending us a tx that is too large which would return with an immediate error. Perhaps we'd want to disconnect from this peer if we see to many of these. I'll open an issue.

Reply:
https://github.com/tendermint/tendermint/issues/6523